Core/include/Acts/Clusterization/InPlaceClusterization.hpp

```diff
@@ -0,0 +1,371 @@
+};
+
+/// @brief base concept for the container that can be used for a cell collection
+template <typename container_t>
+concept SequenceContainer = requires(container_t &cont, unsigned int index) {
+  { cont[index] } -> std::convertible_to<typename container_t::value_type>;
+  { cont.empty() } -> std::convertible_to<bool>;
+  { cont.size() } -> std::convertible_to<std::size_t>;
+} && std::permutable<typename container_t::iterator>;
+
+/// @brief concept of a cell container that can be clustered
+template <typename cell_collection_t>
+concept CellCollection = SequenceContainer<cell_collection_t> &&
+                         CellWithLabel<typename cell_collection_t::value_type>;
+
+/// @brief test whether cells are connected considering common corners and edges.
+/// @param coordinates_diff absolute values of the differences of the cell coordinates
+/// Test whether two cells with the given coordinate differences are connected
+/// either horizontally, vertically, diagonally or are equal.
+template <typename coordinates_t>
+bool isConnectedCommonEdgeOrCorner(const coordinates_t &coordinates_diff) {
+  bool connected = true;
+  for (const auto &a_coordinate_diff : coordinates_diff) {
+    connected &= a_coordinate_diff <= 1;
+  }
+  return connected;
+}
+
+/// @brief test whether cells are connected considering common edges only.
+/// @param coordinates_diff absolute values of the differences of the cell coordinates
+/// Test whether two cells with the given coordinate differences are connected
+/// either horizontally, vertically or are equal.
+template <typename coordinates_t>
+bool isConnectedCommonEdge(const coordinates_t &coordinates_diff) {
+  int connections = 0;
+  for (const auto &a_coordinate_diff : coordinates_diff) {
+    connections += a_coordinate_diff;
+  }
+  return connections <= 1;
+}
+
+/// @brief test whether cells are connected considering common edges or common corners
+enum class EConnectionType { CommonEdgeOrCorner, CommonEdge };
+template <EConnectionType CONNECTION_TYPE, typename coordinates_t>
+bool isConnected(const coordinates_t &coordinates_diff) {
+  if constexpr (CONNECTION_TYPE == EConnectionType::CommonEdgeOrCorner) {
+    return isConnectedCommonEdgeOrCorner(coordinates_diff);
+  } else {
+    return isConnectedCommonEdge(coordinates_diff);
+  }
+}
+
+/// @brief test whether the search for connections can be aborted.
+/// the search is performed in descending order in one coordinate, once the
+/// difference in this coordinate is too large the search can be aborted because
+/// there cannot be any further connections.
+template <typename coordinates_t>
+bool canAbortSearch(const coordinates_t &coordinates_diff,
+                    unsigned int sort_axis_i) {
+  assert(sort_axis_i < coordinates_diff.size());
+  return coordinates_diff[sort_axis_i] > 1;
+}
+
+/// @brief compute the absolute difference of two coordinates
+template <typename coordinate_t>
+auto absDifference(coordinate_t a, coordinate_t b) {
```

Comment:
everywhere else I saw `const coordinate_t &`, here it is passed by value. is there a reason for that?